addons/sale_timesheet/models/account.py

```diff
@@ -93,76 +76,49 @@ def _get_timesheet_billing_values(self, values):
             else:
                 0
         """
-        result = {}
-        if self.env.context.get('create'):  # avoid bad loop
-            return result
-
-        # find the timesheet UoM
-        timesheet_uom = self.product_uom_id
-        if values.get('product_uom_id'):
-            timesheet_uom = self.env['product.uom'].browse(values['product_uom_id'])
-        if not timesheet_uom:  # fallback on default company timesheet UoM
-            timesheet_uom = self.env.user.company_id.project_time_mode_id
-
-        unit_amount = values.get('unit_amount', 0.0) or self.unit_amount
-        billable_type = 'non_billable'
-        revenue = 0.0
-
-        # set the revenue and billable type according to the product and the SO line
-        so_line_id = values.get('so_line') or self.so_line.id
-        so_line = self.env['sale.order.line'].browse(so_line_id).sudo() if so_line_id else self.env['sale.order.line'].browse().sudo()
-        if so_line.product_id.type == 'service':
-            # find the analytic account to convert revenue into its currency
-            account_id = values.get('account_id') or self.account_id.id
-            analytic_account = self.env['account.analytic.account'].browse(account_id)
-            # convert the unit of mesure into hours
-            sale_price_hour = so_line.product_uom._compute_price(so_line.price_unit, timesheet_uom)
-            sale_price = so_line.currency_id.compute(sale_price_hour, analytic_account.currency_id)  # amount from SO should be convert into analytic account currency
-            # calculate the revenue on the timesheet
-            if so_line.product_id.invoice_policy == 'delivery':
-                revenue = analytic_account.currency_id.round(unit_amount * sale_price * (1-(so_line.discount/100)))
-                billable_type = 'billable_time'
-            elif so_line.product_id.invoice_policy == 'order' and so_line.product_id.track_service == 'task':
-                quantity_hour = unit_amount
-                if so_line.product_uom.category_id == timesheet_uom.category_id:
+        for timesheet in self.sudo():
```

Comment:
Perf remark: sudo everything in _timesheet_postprocess and do the postprocess write and _timesheet_compute_theorical_revenue call on this sudoed recordset ?